sparse/client.go

```diff
@@ -264,61 +255,34 @@ func (session *SyncSessionClient) SyncDataInterval(file FileIoProcessor, dataInt
 		/*
 			sync the batch data interval:
 
-			1. send the data range(start and end byte offsets) to ensure the other end
-				has pure data in this range, if so, the other end will ask for checksum
-				and calculate its own at the same time. If not, it will ask for data
-			2. waiting for the reply asking checksum or data
-			3. if the other end asking for checksum, then calculate checksum and send checksum
-			4. if the other end asking for data, then send data
-			5. if the other end asking continue, then data interval is in sync and move on
-				to the next batch interval
+			1. ask server for checksum
+			2. if the other end sending back non-zero length checksum, then calculate local checksum and compare
+			3. if server checksum sent back is zero length or comparison results differs, send data
 		*/
-		log.Debug("syncing data batch interval: ", batchInterval)
-		err := session.send(requestHeader{requestMagic, syncData})
-		if err != nil {
-			log.Debug("encode syncData cmd reply failed: ", err)
-			return err
-		}
-		err = session.send(batchInterval)
-		if err != nil {
-			log.Debug("encode batchInterval: %s failed, err: %s", batchInterval, err)
-			return err
-		}
-
-		var reply replyHeader
-		err = session.receive(&reply)
+		serverCheckSum, err := client.getServerChecksum(batchInterval)
 		if err != nil {
-			log.Debug("decode syncData cmd reply failed: ", err)
+			log.Errorf("client.getServerChecksum(batchInterval) failed, err: %s", err)
 			return err
 		}
+		log.Debugf("server checksum length: %d", len(serverCheckSum))
 
-		if reply.Code == sendChecksum {
-			log.Debug("server reply asking for checksum")
-
-			// calculate checksum for the data batch interval
+		serverNeedData := true
+		if len(serverCheckSum) != 0 {
```

Comment:
Do you have test case for this `if` branch? Say, got checksum, then found it's different vs found it's the same.